src/model/rpc.js

```diff
@@ -1,141 +1,162 @@
+          case clpPacket.TYPE_ACK:
+          case clpPacket.TYPE_RESPONSE:
+            resolve(data)
             break
-          case Clp.TYPE_RESPONSE:
-            resolve(Clp.deserializeResponse(message))
-            break
-          case Clp.TYPE_ERROR:
-            reject(new Error(JSON.stringify(Clp.deserializeResponse(message))))
+
+          case clpPacket.TYPE_ERROR:
+            reject(new Error(JSON.stringify(data)))
             break
+
+          default:
+            throw new Error('Unkown CLP packet type', data)
         }
       })
     })
 
     const timeout = new Promise((resolve, reject) =>
-      setTimeout(() => reject(new Error(id + ' timed out')), DEFAULT_TIMEOUT))
+      setTimeout(() => {
+        reject(new Error(id + ' timed out'))
+      }, DEFAULT_TIMEOUT))
 
     return Promise.race([
       response,
       timeout
     ])
   }
 
-  async prepare ({ id, amount, executionCondition, expiresAt, protocolData }) {
-    const requestId = uuid()
-    const prepareRequest = Clp.serializePrepare({
-        id, amount, executionCondition, expiresAt
-      }, requestId, protocolData)
+  async prepare ({id, amount, executionCondition, expiresAt}, protocolData) {
+    const requestId = await _requestId()
+    const prepareRequest = clpPacket.serializePrepare({
+      transferId: id,
+      amount,
+      executionCondition,
+      expiresAt
+    }, requestId, protocolData)
 
-    return _call(requestId, prepareRequest)
+    return this._call(requestId, prepareRequest)
   }
 
-  async fulfill ({ id, fulfillment, protocolData }) {
-    const requestId = uuid()
-    const fulfillRequest = Clp.serializeFulfill({
-        id, fulfillment
-      }, requestId, protocolData)
+  async fulfill ({id, fulfillment}, protocolData) {
+    const requestId = await _requestId()
+    const fulfillRequest = clpPacket.serializeFulfill({
+      transferId: id,
+      fulfillment
+    }, requestId, protocolData)
 
-    return _call(requestId, fulfillRequest)
+    return this._call(requestId, fulfillRequest)
   }
 
-  async reject ({ id, reason, protocolData }) {
-    const requestId = uuid()
-    const rejectRequest = Clp.serializeReject({
-        id, reason
-      }, requestId, protocolData)
+  async reject ({id, rejectionReason}, protocolData) {
+    const requestId = await _requestId()
+    const rejectRequest = clpPacket.serializeReject({
+      transferId: id,
+      rejectionReason
+    }, requestId, protocolData)
 
-    return _call(requestId, rejectRequest)
+    return this._call(requestId, rejectRequest)
   }
 
   async message ({ protocolData }) {
-    const requestId = uuid()
-    const messageRequest = Clp.serializeReject(requestId, protocolData)
+    const requestId = await _requestId()
+    const messageRequest = clpPacket.serializeMessage(requestId, protocolData)
 
-    return _call(requestId, messageRequest)
+    return this._call(requestId, messageRequest)
   }
 
   async _connect () {
```

Comment:
authentication to the other side should be done in here, with a bearer token passed into the constructor

Reply:
> with a bearer token passed into the constructor

If you mean passing the bearer token to the WebSocket constructor like `new WebSocket(uri + 'token=1234')`, then I disagree. This makes the token end up in the URL, which is not good security practice.

After having a very similar challenge in fb-ledger, I think the cleanest solution is to send the bearer token as first message once the websocket is open.